src/setup_payload/tests/TestAdditionalDataPayload.cpp

```diff
@@ -0,0 +1,283 @@
+#include <lib/support/CHIPPlatformMemory.h>
+#include <setup_payload/AdditionalDataPayloadGenerator.h>
+#include <setup_payload/AdditionalDataPayloadParser.h>
+#include <setup_payload/SetupPayload.h>
+#include <system/SystemPacketBuffer.h>
+
+#include <lib/support/UnitTestRegistration.h>
+#include <lib/support/verhoeff/Verhoeff.h>
+
+using namespace chip;
+
+namespace {
+
+constexpr char kSerialNumber[]                                           = "89051AAZZ236";
+constexpr char kAdditionalDataPayloadWithoutRotatingDeviceId[]           = "1518";
+constexpr char kAdditionalDataPayloadWithRotatingDeviceId[]              = "153000120A001998AB7130E38B7E9A401CFE9F7B79AF18";
+constexpr char kAdditionalDataPayloadWithInvalidRotatingDeviceIdLength[] = "153000FF0A001998AB7130E38B7E9A401CFE9F7B79AF18";
+constexpr char kAdditionalDataPayloadWithLongRotatingDeviceId[]          = "153000130A00191998AB7130E38B7E9A401CFE9F7B79AF18";
+constexpr char kAdditionalDataPayloadWithRotatingDeviceIdAndMaxLifetimeCounter[] = "15300012FFFFFC1670A9F9666D1C4587FCBC4811549018";
+constexpr char kRotatingDeviceId[]                                               = "0A001998AB7130E38B7E9A401CFE9F7B79AF";
+constexpr uint16_t kLifetimeCounter                                              = 10;
+constexpr uint16_t kAdditionalDataPayloadLength                                  = 51;
+
+CHIP_ERROR GenerateAdditionalDataPayload(nlTestSuite * inSuite, uint16_t lifetimeCounter, const char * serialNumberBuffer,
+                                         size_t serialNumberBufferSize, BitFlags<AdditionalDataFields> additionalDataFields,
+                                         char * additionalDataPayloadOutput)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::System::PacketBufferHandle bufferHandle;
+
+    err = AdditionalDataPayloadGenerator().generateAdditionalDataPayload(
+        lifetimeCounter, serialNumberBuffer, serialNumberBufferSize, bufferHandle, additionalDataFields);
+    if (err == CHIP_NO_ERROR)
+    {
+        NL_TEST_ASSERT(inSuite, !bufferHandle.IsNull());
+    }
+    else
+    {
+        return err;
+    }
+    char output[kAdditionalDataPayloadLength];
+    err = chip::Encoding::BytesToUppercaseHexString(bufferHandle->Start(), bufferHandle->DataLength(), output, ArraySize(output));
+
+    if (err == CHIP_NO_ERROR)
+    {
+        memmove(additionalDataPayloadOutput, output, kAdditionalDataPayloadLength);
+    }
+    return err;
+}
+
+CHIP_ERROR ParseAdditionalDataPayload(const char * additionalDataPayload,
+                                      chip::SetupPayloadData::AdditionalDataPayload & outPayload)
+{
+    std::vector<uint8_t> payloadData;
+    std::string payloadString(additionalDataPayload);
+
+    // Decode input payload
+    size_t len = payloadString.length();
+
+    for (size_t i = 0; i < len; i += 2)
+    {
+        auto str  = payloadString.substr(i, 2);
+        uint8_t x = (uint8_t) stoi(str, 0, 16);
+        payloadData.push_back(x);
+    }
```

Comment:
Please use `HexToBytes()` from `BytesToHex.h` instead of writing your own...

Also, there is no need to use a vector here...

Reply:
good point, updated